cpp/examples/arrow/execution_plan_documentation_examples.cc

```diff
@@ -0,0 +1,1160 @@
+
+    auto partitioning =
+    std::make_shared<arrow::dataset::HivePartitioning>(partition_schema);
+    // We'll write Parquet files.
+    auto format = std::make_shared<arrow::dataset::ParquetFileFormat>();
+
+    arrow::dataset::FileSystemDatasetWriteOptions write_options;
+    write_options.file_write_options = format->DefaultWriteOptions();
+    write_options.filesystem = filesystem;
+    write_options.base_dir = base_path;
+    write_options.partitioning = partitioning;
+    write_options.basename_template = "part{i}.parquet";
+
+    arrow::dataset::WriteNodeOptions write_node_options {write_options,
+    dataset->schema()};
+
+    PRINT_LINE("Write Options created");
+
+    ARROW_ASSIGN_OR_RAISE(cp::ExecNode *wr, cp::MakeExecNode("write", plan.get(),
+    {scan}, write_node_options));
+
+    ABORT_ON_FAILURE(wr->Validate());
+    ABORT_ON_FAILURE(plan->Validate());
+    PRINT_LINE("Execution Plan Created : " << plan->ToString());
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
+    plan->finished().Wait();
+    return arrow::Status::OK();
+}
+
+arrow::Status SourceUnionSinkExample() {
+    auto basic_data = MakeBasicBatches();
+
+    cp::ExecContext exec_context(arrow::default_memory_pool(),
+                                 ::arrow::internal::GetCpuThreadPool());
+
+    std::shared_ptr<cp::ExecPlan> plan =
+    cp::ExecPlan::Make(&exec_context).ValueOrDie();
+    arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+    cp::Declaration union_node{"union", cp::ExecNodeOptions{}};
+    cp::Declaration lhs{"source",
+                  cp::SourceNodeOptions{basic_data.schema,
+                                    basic_data.gen(/*parallel=*/false)}};
+    lhs.label = "lhs";
+    cp::Declaration rhs{"source",
+                    cp::SourceNodeOptions{basic_data.schema,
+                                      basic_data.gen(/*parallel=*/false)}};
+    rhs.label = "rhs";
+    union_node.inputs.emplace_back(lhs);
+    union_node.inputs.emplace_back(rhs);
+
+    cp::CountOptions options(cp::CountOptions::ONLY_VALID);
+    ARROW_ASSIGN_OR_RAISE(auto declr,
+    cp::Declaration::Sequence(
+            {
+                union_node,
+                {"aggregate", cp::AggregateNodeOptions{
+                  /*aggregates=*/{{"count", &options}},
+                  /*targets=*/{"a"},
+                  /*names=*/{"count(a)"},
+                  /*keys=*/{}}},
+                {"sink", cp::SinkNodeOptions{&sink_gen}},
+            })
+            .AddToPlan(plan.get()));
+
+    ABORT_ON_FAILURE(declr->Validate());
+
+    ABORT_ON_FAILURE(plan->Validate());
+    std::shared_ptr<arrow::RecordBatchReader> sink_reader = cp::MakeGeneratorReader(
+        arrow::schema({arrow::field("count(a)", arrow::int32())}),
+        std::move(sink_gen),
+        exec_context.memory_pool());
+
+    // // // start the ExecPlan
+    ABORT_ON_FAILURE(plan->StartProducing());
```

Comment:
Why abort here instead of `ARROW_RETURN_NOT_OK`?

Reply:
Yes, it suits better. Replaced.